### PR TITLE
conversation: use standard message format to pass history

### PR DESCRIPTION
**Pull request by mamuqsit**
## Introduction

ASHABot was passing conversation history to the LLM as a flat serialized string (`query1: ... answer1: ...`) which caused model to poorly resolve elliptical follow-up queries (particularly "Yes" and "No" responses), often misclassifying them as `small_talk` or `incomprehensible`.

PR replaces homegrown history format with the standard role-based message format [that LLMs are natively trained on](https://learn.microsoft.com/en-us/azure/ai-foundry/openai/how-to/chatgpt?view=foundry-classic&tabs=python-secure#:~:text=OpenAI%20trained%20chat%20completion%20models%20to%20accept%20input%20formatted%20as%20a%20conversation) so the model is given a proper conversational context for query rewriting and classification.
```diff
- conversation_history.append(f"query{i}: {question} answer{i}: {answer}")
- i += 1
+ conversation_history.append({"role": "user", "content": question})
+ conversation_history.append({"role": "assistant", "content": answer})
```

## Changes
- Replace serialized string history with a structured `[{"role": "user", ...}, {"role": "assistant", ...}]` list passed directly into the LLM messages array alongside the system prompt
- Update `query_rewrite` example to use the new role-based format
- Add `QUESTION_REWRITTEN` in `constants.py` to separately store rewritten form of query. `QUESTION` now stores original source text

### CODE REVIEW

**Review by a4i-architect — COMMENTED** [review]
### Summary of Review

#### Key Changes & Improvements:
- **Timestamp update**:** Transition from `int(time.time())` to `datetime.now(timezone.utc)` in chat.py and background job files is good for accurate timestamps.
- **Question fields:** Renaming or adding `QUESTION_REWRITTEN` demonstrates intent to distinguish original vs. rewritten questions, which is positive.
- **Conversation history:** In user handlers, conversation history is now built as a list of role-content dicts, aligning with typical LLM prompts. However, legacy string concatenation is still used in `send.py`, risking inconsistency.
- **XML parsing:** The regex-based XML extraction is fragile; consider a more robust parser like `xml.etree.ElementTree`.

#### Issues & Missing Aspects:
- **Inconsistent `constants.QUESTION_REWRITTEN`:** Defined in `chat.py` and `constants.py`, but sometimes it's not populated (e.g., in `chat.py` the context refers to `QUESTION_REWRITTEN`, but its value isn't always provided). Need to ensure consistent usage.
- **Redundant code:** The conversation history in `user_flow_handlers/process.py` is built twice. One version is a list of dicts, another is legacy string format. Prefer the structured list; remove legacy format altogether.
- **Type hints:** Slight inconsistencies, e.g., `_create_conversation_history()` returning a `list[dict[str, str]]`, which improves clarity.
- **Unnecessary concatenations:** In `process.py`, `conversation_history_str` is created but not used. Instead, the list form is used for prompts, which is good. Remove unused string joins.
- **Unclean handling of timestamps:** Transition from `int(time.time())` to datetime objects is good; make sure all downstream processes handle datetime objects correctly.
- **Implicit assumptions:** In `chat.py`, `datetime.now()` might be naive if not always consistent with timezone-aware objects used elsewhere.
- **Design complexity:** The code attempts to support multiple languages, translation, and rewriting prompts within the LLM chain, which is complex. Ensure thorough testing of prompt assembly and parsing.

### Recommendations:
- Standardize `conversation_history` as list[dict], removing legacy string concatenation.
- Consolidate prompt construction to avoid duplications.
- Use `xml.etree.ElementTree` for XML parsing robustness.
- Confirm all timestamps are timezone-aware wherever used.
- Add type hints explicitly and verify all function signatures for clarity.
- Increase commenting on complex logic (e.g., conversation chaining, prompt formatting).

### Score: **78/100**  
- The core improvements (timestamps, conversation structure) are solid.
- Some redundancy, inconsistent handling, and fragility in XML parsing prevent a higher score.
- With refinements, the code can be substantially cleaner and more maintainable.

File: byoeb-v1/byoeb/byoeb/chat_app/bot_config.json [hunk 1]
```diff
@@ -12,19 +12,16 @@
...
+                "te": "2. Translate the source Telugu query (query_src) to English (query_en), using the below heuristics: \n\n2a. Identify and REPLACE all terms in query_src that have a mapping defined in json_dictionary_te. \n\n<json_dictionary_te> \n\n{     \n\n    \"ఆశ\": \"ASHA (Accredited Social Health Activist)\", \n\n  } \n\n</json_dictionary_te> \n\n2b. query_src has been written by an ASHA with limited literacy. Try to make sense of it as much as possible while rendering a faithful translation. Typically, ASHAs use a combination of Telugu and English, at times by even combining Latin and Telugu scripts. Ignore grammatical and spelling errors."
             },
-            "user_prompt": "<query_src><QUERY></query_src>\n<conversation_history><CONVERSATION_HISTORY></conversation_history>"
```
Comment (N1kunj1998) [inline byoeb-v1/byoeb/byoeb/chat_app/bot_config.json:27]: User Prompt is still being referred in  `byoeb-v1/byoeb/byoeb/services/admin/message_process.py` at line 80: `template_user_prompt = bot_config["llm_response"]["translation_and_rewrite_prompts"]["user_prompt"]`

we should make change in that file to follow new structure
```
user_prompt = question  # raw source text, no template
conversation_history = get_conversation_history(user_id, history_length)  # returns role-based list
augmented_prompts = [
    {"role": "system", "content": system_prompt},
    *conversation_history,
    {"role": "user", "content": user_prompt},
]
```

File: byoeb-v1/byoeb/byoeb/chat_app/bot_config.json [hunk 2]
```diff
@@ -12,19 +12,16 @@
...
... Else, use past messages to modify query_en and create query_en_addcontext. query_en_addcontext should be **SELF-CONTAINED**, i.e., it should include all necessary context and details so that it can be answered **COMPLETELY** and **APPROPRIATELY** without referencing previous queries. Use the example below as reference.\n\n<eg_addcontext>\n<messages>\n[\n    {\"role\": \"user\", \"content\": \"Hello\"},\n    {\"role\": \"assistant\", \"content\": \"Hello!\"},\n    {\"role\": \"user\", \"content\": \"what are the side effects of the antara injection?\"},\n    {\"role\": \"assistant\", \"content\": \"A woman who receives an Antara injection can experience changes in menstrual bleeding patterns. For most women, the monthly bleeding becomes very light, or there may be spotting or irregular bleeding. In about half of the women, periods stop after a year of use. Monthly bleeding resumes when a woman stops getting injections.\"},\n    {\"role\": \"user\", \"content\", \"tell me about the side effects\"}\n]\n</messages>\n\n<query_en_addcontext>Can you tell me more about the side effects of the Antara injection, beyond changes in menstrual bleeding?</query_en_addcontext>\n</eg_addcontext>",
```
Comment (N1kunj1998) [inline byoeb-v1/byoeb/byoeb/chat_app/bot_config.json]: there is a typo in this line
```
{"role": "user", "content", "tell me about the side effects"}
                           ^  should be ':'
```

fix :
```
{"role": "user", "content": "tell me about the side effects"}
```

File: byoeb-v1/byoeb/byoeb/chat_app/bot_config.json [hunk 3]
```diff
@@ -12,19 +12,16 @@
...
...c may contain typing mistakes such as spelling errors or acronyms written in lowercase or mixed-case. To correct this, first identify and **REPLACE** all terms in query_src that have a corresponding mapping in json_dictionary. Next, capitalize any remaining lowercase or mixed-case acronyms (in Latin script) that are not present in json_dictionary.\n\n<json_dictionary>\n\n{\n\n    \"bipi\": \"BP (blood pressure)\",\n\n    \"pen travelion\": \"pentavalent vaccine\",\n\n    \"Abha\": \"ABHA (Ayushman Bharat Health Account) ID\",\n\n    \"abha\": \"ABHA (Ayushman Bharat Health Account) ID\",\n\n    \"ABDM\": \"ABDM (Ayushman Bharat Digital Mission)\",\n\n    \"abdm\": \"ABDM (Ayushman Bharat Digital Mission)\",\n\n    \"PNC\": \"PNC (Post-Natal Checkup)\",\n\n    \"pnc\": \"PNC (Post-Natal Checkup)\",\n\n    \"FIPV\": \"FIPV (Fractional-dose inactivated poliovirus vaccine)\",\n\n    \"fipv\": \"FIPV (Fractional-dose inactivated poliovirus vaccine)\",\n\n    \"HBNC\": \"HBNC (Home-Based Newborn Care)\",\n\n    \"hbnc\": \"HBNC (Home-Based Newborn Care)\",\n\n    \"HPNC\": \"HBNC (Home-Based Newborn Care)\",\n\n    \"hpnc\": \"HBNC (Home-Based Newborn Care)\"\n\n}\n\n</json_dictionary>",
```
Comment (N1kunj1998) [inline byoeb-v1/byoeb/byoeb/chat_app/bot_config.json]: since we have removed query_src xml tag, I think we can re-write this prompt without

- `query_src may contain typing mistakes`

File: byoeb-v1/byoeb/byoeb/chat_app/bot_config.json [hunk 4]
```diff
@@ -12,19 +12,16 @@
...
...             "hi": "2. Translate the source Hindi query (query_src) to English (query_en), using the below heuristics:\n\n2a. Identify and **REPLACE** all terms in query_src that have a mapping defined in json_dictionary_hindi.\n\n<json_dictionary_hindi>\n\n{     \n\n    \"भजन\": \"वजन\",  \n\n    \"अन्ना\": \"अंतरा\",  \n\n    \"bhajan\": \"vajan\",  \n\n    \"anna\": \"antara\",  \n\n    \"आशा\": \"ASHA (Accredited Social Health Activist)\" \n\n}\n\n</json_dictionary_hindi>\n\n2b. query_src has been written by an ASHA with limited literacy. Try to make sense of it as much as possible while rendering a faithful translation. Typically, ASHAs use Hinglish (a combination of Hindi and English), at times by even combining Latin and Devanagari scripts. Ignore grammatical and spelling errors.\n<eg_translation>\n\n<eg1><query_src>Kamer me drard ho rha h</query_src><query_en>I am having back pain</query_en></eg1>\n\n<eg2><query_src>Yishu ka growth chart कैसे भरा jata hai</query_src><query_en>How to fill a baby’s growth chart</query_en></eg2>\n\n<eg3><query_src>ABDM kab shuru hui</query_src><query_en>When did the Ayushman Bharat Digital Mission start?</query_en></eg3>\n\n</eg_translation>",
```
Comment (N1kunj1998) [inline byoeb-v1/byoeb/byoeb/chat_app/bot_config.json]: since we have removed query_src xml tag, I think we can re-write this prompt without

- `Translate the source Hindi query (query_src) to English (query_en)`

File: byoeb-v1/byoeb/byoeb/services/chat/message_handlers/user_flow_handlers/process.py [hunk 5]
```diff
@@ -179,10 +171,14 @@ async def handle_process_message_workflow(
         else:
             # Fallback for AUDIO_IDK or other cases
             message.message_context.message_english_text = message.message_context.message_source_text
+
+        conv_history = self._create_conversation_history(message.user.last_conversations)
```
Comment (N1kunj1998) [inline byoeb-v1/byoeb/byoeb/services/chat/message_handlers/user_flow_handlers/process.py]: We are doing two calls for conversation history for each message, should we not combine them, other call is present at `line #87` 

Fix:
```
# In llm_translation_and_query_rewritting, return the history too, or store on the first call
conversation_history = self._create_conversation_history(messages.user.last_conversations)
augmented_prompts = self.__augment(system_prompt, user_prompt, conversation_history)
...

# In handle_process_message_workflow, reuse result
conversation_history = ...  # already computed above
conv_history_legacy = []
for i in range(0, len(conversation_history) - 1, 2):
    user_turn = conversation_history[i]['content']
    assistant_turn = conversation_history[i + 1]['content']
    conv_history_legacy.append(f"query{(i//2)+1}: {user_turn} answer{(i//2)+1}: {assistant_turn}")
```

**Comment by N1kunj1998** [comment]
I have tested code in locally via mcp, functionality is working fine and addresses the PR issue, just few small tweaks to do before merging.

**Review by a4i-architect — COMMENTED** [review]
Summary of Key Issues and Recommendations:

1. **DateTime Handling**:  
   - Replace all `datetime.now(timezone.utc)` with a central utility function or context-aware provider to ensure consistent timestamping.

2. **Code Duplication & Redundancy**:  
   - Multiple places construct conversation histories; unify into a single method (e.g., `_create_conversation_history`) to ensure consistent format and reduce repetition.

3. **Type Annotations & Typing Consistency**:  
   - In `byoeb/services/chat/message_handlers/user_flow_handlers/process.py`, the method signature should specify exact list typing `list[dict[str, str]]` instead of the generic `list`.

4. **XML Parsing Logic**:  
   - Parsing response using regex is fragile; consider using an XML parser like `xml.etree.ElementTree` for robustness.

5. **Code Immutability & Structure**:
   - In several files, usage of mutable lists vs. tuples or namedtuples can improve clarity and reduce bugs, especially around conversation history.

6. **Response Handling & Formatting**:
   - Consistency checks: ensure all functions return and parse XML responses in the expected format, especially after updates.
   
7. **Message Context & Constants**:
   - The dual use of `QUESTION` and `QUESTION_REWRITTEN` is good for clarity but ensure all code consistently updates these fields and documents their usage.

8. **Design Pattern**:
   - The pattern of embedding conversation history and prompt generation is good; however, the repeated appending of histories suggests a potential for a dedicated conversation state manager.

9. **Testing & Validation**:
   - Add automated tests for timestamp correctness, conversation history formatting, and XML parsing stability.

**Overall Score**: 75/100  
- Well-structured but some redundancy, weak parsing robustness, and minor type improvements needed.  
- Acceptance contingent on addressing the XML parsing and unifying conversation history creation.

**References**:  
- [Effective timestamping in distributed systems](https://martinfowler.com/articles/elapsed-time.html)  
- [Python XML parsing best practices](https://docs.python.org/3/library/xml.etree.elementtree.html)  
- [Type annotations in Python](https://docs.python.org/3/library/typing.html)

File: byoeb-v1/byoeb/byoeb/chat_app/bot_config.json (hunk 1 above)
Comment (mamuqsit) [reply byoeb-v1/byoeb/byoeb/chat_app/bot_config.json:27]: Addressed in https://github.com/A4i-tech/byoeb/pull/147/commits/f53713794abfdb7cf6693ada322f36365081d367

File: byoeb-v1/byoeb/byoeb/chat_app/bot_config.json (hunk 2 above)
Comment (mamuqsit) [reply byoeb-v1/byoeb/byoeb/chat_app/bot_config.json]: Addressed in https://github.com/A4i-tech/byoeb/pull/147/commits/a9c5930866dfbf7365cf27e2e5d5ea6b30e78bae

File: byoeb-v1/byoeb/byoeb/chat_app/bot_config.json (hunk 3 above)
Comment (mamuqsit) [reply byoeb-v1/byoeb/byoeb/chat_app/bot_config.json]: Addressed in https://github.com/A4i-tech/byoeb/pull/147/commits/d0be45a393a6a5e13e3992e288d806b834ba45d4

File: byoeb-v1/byoeb/byoeb/chat_app/bot_config.json (hunk 4 above)
Comment (mamuqsit) [reply byoeb-v1/byoeb/byoeb/chat_app/bot_config.json]: Addressed in https://github.com/A4i-tech/byoeb/pull/147/commits/d0be45a393a6a5e13e3992e288d806b834ba45d4

File: byoeb-v1/byoeb/byoeb/services/chat/message_handlers/user_flow_handlers/process.py (hunk 5 above)
Comment (mamuqsit) [reply byoeb-v1/byoeb/byoeb/services/chat/message_handlers/user_flow_handlers/process.py]: Addressed in https://github.com/A4i-tech/byoeb/pull/147/commits/edf2502680df32e74a7814cc71bffb226a6eb02d

**Review by N1kunj1998 — APPROVED** [review]
Changes look good, Test via MCP in local.

File: byoeb-v1/byoeb/byoeb/services/admin/message_process.py [hunk 6]
```diff
@@ -76,13 +75,8 @@ def parse_xml_with_regex(xml_string: str):
...
         return extracted_data[QUERY_EN], extracted_data[QUERY_EN_ADDCONTEXT], extracted_data[QUERY_TYPE]
-    conversation_history_str = ", ".join(conversation_history)
-    template_user_prompt = bot_config["llm_response"]["translation_and_rewrite_prompts"]["user_prompt"]
-    user_prompt = template_user_prompt.replace("<QUERY>", question).replace("<CONVERSATION_HISTORY>", conversation_history_str)
-    augmented_prompts = [
-        {"role": "system", "content": system_prompt},
-        {"role": "user", "content": user_prompt}
-    ]
+    conversation_history = byoeb_user_process._create_conversation_history(conversation_history)
+    augmented_prompts = byoeb_user_process.__augment(system_prompt, question, conversation_history)
```
Comment (theainerd) [inline byoeb-v1/byoeb/byoeb/services/admin/message_process.py]: Blocking: this calls a double-underscore method across modules (byoeb_user_process.__augment), which is name-mangled and not accessible as __augment externally. This will raise AttributeError at runtime in /experiment flow. Please expose a public helper (e.g. _augment or build_augmented_prompts) and call that instead.

File: byoeb-v1/byoeb/byoeb/services/admin/message_process.py [hunk 7]
```diff
@@ -76,13 +75,8 @@ def parse_xml_with_regex(xml_string: str):
...
 
         return extracted_data[QUERY_EN], extracted_data[QUERY_EN_ADDCONTEXT], extracted_data[QUERY_TYPE]
-    conversation_history_str = ", ".join(conversation_history)
-    template_user_prompt = bot_config["llm_response"]["translation_and_rewrite_prompts"]["user_prompt"]
-    user_prompt = template_user_prompt.replace("<QUERY>", question).replace("<CONVERSATION_HISTORY>", conversation_history_str)
-    augmented_prompts = [
-        {"role": "system", "content": system_prompt},
-        {"role": "user", "content": user_prompt}
-    ]
+    conversation_history = byoeb_user_process._create_conversation_history(conversation_history)
```
Comment (theainerd) [inline byoeb-v1/byoeb/byoeb/services/admin/message_process.py]: Behavior regression: conversation_history is already in standard [{"role","content"}] format from get_conversation_history(), but we pass it again into _create_conversation_history(), which expects timestamp/question/answer records. This drops all history and sends empty context to rewrite LLM. Please remove this second transformation or pass raw last_conversations shape.

**Review by a4i-architect — COMMENTED** [review]
**Overall Summary:**

The primary modifications introduce timestamp updates using `datetime.now(timezone.utc)` instead of `time.time()`, improving time zone awareness and datetime handling. Additional fields like `QUESTION_REWRITTEN` are added to maintain conversation context and answer rewriting.

**Key Issues & Recommendations:**

1. **Inconsistent Timestamp Handling:**
   - Replacing `time.time()` with `datetime` objects improves clarity but requires consistent parsing downstream. Ensure all consumers expect `datetime` objects, not strings or integers.
   - Consider serializing timestamps uniformly (e.g., ISO format) for storage and retrieval.

2. **Redundant or Confusing Code:**
   - In `chat.py`, the timestamp is assigned but not explicitly serialized for DB or logs. Confirm storage consistency.
   - In `message_handlers/user_flow_handlers/process.py`, `build_augmented_prompts` is added but should be verified for correct integration with existing functions.

3. **Unused or Duplicate Fields:**
   - `QUESTION_REWRITTEN` field is added multiple times; ensure all code paths handle it correctly.
   - The conversation history transformation at the end of `process.py` is converted from structured lists to legacy string format. Maintain both if older clients depend on it, or document the dual format usage.

4. **Test Coverage:**
   - The integration test effectively verifies conversation history captures multiple languages; ensure tests also cover edge cases like missing timestamps or malformed data.

5. **Design & Complexity:**
   - The refactoring to `build_augmented_prompts` encapsulates prompt assembly but should be reviewed for correctness across all callers.
   - The use of `@retry` is appropriate but could be further evaluated for robustness.

**Score: 80/100**

**Acceptance Justification:**

The changes improve datetime handling and conversation context management but require careful validation for timestamp serialization and downstream compatibility. Minor adjustments to serialization and documentation are recommended before full deployment.

File: byoeb-v1/byoeb/byoeb/services/admin/message_process.py (hunk 6 above)
Comment (mamuqsit) [reply byoeb-v1/byoeb/byoeb/services/admin/message_process.py]: Addressed in https://github.com/A4i-tech/byoeb/pull/147/changes/fb193093157b2f97d2f6f1c79217d10f223ff7b3

File: byoeb-v1/byoeb/byoeb/services/admin/message_process.py (hunk 7 above)
Comment (mamuqsit) [reply byoeb-v1/byoeb/byoeb/services/admin/message_process.py]: Missed this cause I didn't test out the /experiment endpoint, addressed in https://github.com/A4i-tech/byoeb/pull/147/changes/2cbdc7d83f247bb50e30befa5d15436086a1e1ec

**Review by theainerd — APPROVED** [review]
Approved with one follow-up: preserve prior conversation_history for onboarding and AUDIO_IDK paths when LLM rewriting is skipped.

**Comment by N1kunj1998** [comment]
@mamuqsit, integration tests for embedding cache are failing in this PR, can you check this

[Ideate conversation flow - Integration test failing.txt](https://github.com/user-attachments/files/26074654/Ideate.conversation.flow.-.Integration.test.failing.txt)

```
===================================== short test summary info ===================================== 
FAILED tests/integration/utils/test_embedding_cache.py::test_repeated_query_hits_cache - AssertionError: assert not True
====================== 1 failed, 34 passed, 17 warnings in 499.66s (0:08:19) ======================
```

**Review by a4i-architect — COMMENTED** [review]
Below is your complete review following all enforced team standards and scoring rules.

---

## High‑Level Summary

The shift to role‑based LLM conversation history is broadly correct and aligns with expected model behavior. However, several regressions introduce **silent failures**, **runtime type inconsistencies**, **logging policy violations**, and **propagation gaps**. A few areas introduce **CRITICAL security issues** (logging sensitive conversational content) and **HIGH correctness bugs** (timestamps, history reconstruction, API typing). All issues below cite file + function and include concrete fixes.

---

# Issues

---

## CRITICAL ISSUES (‑8 each)

### 1. **PII / Sensitive Query Content Logged in Plaintext**
**File:** `services/chat/message_handlers/user_flow_handlers/process.py`, function `handle_process_message_workflow`  
**Issue:** The logging call inside the token‑logging section records the **full query text** (unmasked) in AppInsights, violating the “no PII/audio/content in logs” rule.  
**Fix:** Replace text content with masked variants (`utils.mask_text`) or remove content entirely from logs.

---

### 2. **Structured logging bypassed using print statements**
**File:** `tests/integration/test_message_history.py`  
**Issue:** Uses `print("Environment variables are missing")`, violating structured logging requirement.  
**Fix:** Replace with structured logging (e.g., `logger.error("Missing environment variables", extra={...})`).

---

## HIGH ISSUES (‑5 each)

### 3. **Incorrect TIMESTAMP type stored as datetime instead of str**
**File:** `apis/chat.py`, function `asha_chat`  
**Issue:** Previously stored UNIX timestamps as `str(int(time.time()))`. Now storing raw `datetime` breaks downstream consumers that expect a string.  
**Fix:** Convert to ISO string (`datetime.now(timezone.utc).isoformat()`).

---

### 4. **Conversation history reconstruction generates malformed legacy strings**
**File:** `services/chat/message_handlers/user_flow_handlers/process.py`, function `handle_process_message_workflow`  
**Issue:** Reconstructing legacy history assumes every pair is `[user, assistant]`. If list length is odd (possible when upstream data is incomplete) → crashes or misaligns.  
**Fix:** Add guard: skip chunks that do not have exactly two valid entries.

---

### 5. **History creation silently skips invalid timestamps without logging**
**File:** `_create_conversation_history` (same file above)  
**Issue:** Violates error‑handling standards: errors must be logged with context before skipping.  
**Fix:** Log a WARNING including conversation id and invalid timestamp type.

---

### 6. **llm_translation_and_query_rewritting return signature changed but callers still expect previous behavior**
**File:** `services/admin/message_process.py`, function `llm_translation_and_query_rewritting`  
**Issue:** Now returns `(query_en, query_en_addcontext, query_type)` only; previous callers expected four fields. Silent mismatch causes incorrect unpacking in some paths.  
**Fix:** Restore full return signature or update all callsites consistently.

---

### 7. **Tests access response.additional_info incorrectly**
**File:** `tests/integration/test_message_history.py`, `test_conversation_history_is_captured`  
**Issue:** `for k, v in response.data.additional_info` assumes dict is iterable as key/value tuples; iteration yields keys only. Test will crash silently.  
**Fix:** Replace with `response.data.additional_info["Conversation history"]`.

---

## MEDIUM ISSUES (‑3 each)

### 8. **Regex parsing lacks error handling**
**File:** `services/admin/message_process.py`, function `parse_xml_with_regex`  
**Issue:** No error logging when XML fields missing. Violates error‑handling rules.  
**Fix:** Log an ERROR with the raw response when regex fails before returning None fields.

---

### 9. **Tenacity retry decorator missing error boundary**
**File:** `process.py`, `llm_translation_and_query_rewritting`  
**Issue:** Removed `RetryError` import but now callers no longer catch it, causing unhandled retry exhaustion.  
**Fix:** Add explicit handling for retry exhaustion and log the failure before returning fallback.

---

### 10. **Mixing synchronous heavy operations in async workflow**
**File:** `handle_process_message_workflow`  
**Issue:** CPU‑heavy string operations (history flattening, regex parsing) run inside async handler, potentially blocking event loop.  
**Fix:** Offload heavy parsing to a thread executor via `loop.run_in_executor`.

---

## LOW ISSUES (‑1 each)

### 11. **Type hints missing (`list[dict[str,str]]`) violate STRICT TYPING**
**File:** `build_augmented_prompts`, `create_conversation_history`  
**Fix:** Use full type aliases: `list[dict[str, str]]`.

---

### 12. **Config fragments still contain large inline magic values**
**File:** `bot_config.json`  
**Issue:** Huge string blocks → violates architecture rule requiring centralized configuration values.  
**Fix:** Move large prompt blocks into separate versioned prompt files.

---

### 13. **Unused imports**
**Files:**  
- `services/admin/message_process.py` (`AzureVectorStore` previously removed but others still unused)  
**Fix:** Remove unused imports.

---

# Score Calculation

CRITICAL: 2 × 8 = 16  
HIGH: 4 × 5 = 20  
MEDIUM: 3 × 3 = 9  
LOW: 3 × 1 = 3  

Total deductions = **48**

**Final Score = 100 − 48 = 52**

---

## Optional References
- Structured logging patterns: https://docs.python.org/3/howto/logging.html  
- Secure logging guidelines: https://owasp.org/www-community/Logging_Cheat_Sheet
